tensorflow/examples/how_tos/reading_data/fully_connected_reader.py

```diff
@@ -45,11 +45,9 @@
 VALIDATION_FILE = 'validation.tfrecords'
 
 
-def read_and_decode(filename_queue):
-  reader = tf.TFRecordReader()
-  _, serialized_example = reader.read(filename_queue)
-  features = tf.parse_single_example(
-      serialized_example,
+def decode(serialized_example):
+  features = tf.parse_example(
```

Comment:
IMO, this would be a bit simpler as `features = tf.parse_single_example(serialized_example, ...)` rather than using the batch version and explicitly adding a dimension to `serialized_example`. You wouldn't need the `tf.reshape(image, [-1])` below either (or the later reshape of the label value), and the comment about "scalar string tensor" would remain correct (in the current PR, it's a length-1 vector of strings).

Did you have a reason to prefer switching it here?

Reply:
I stumbled upon #390 . I wasn't sure if it was still relevant so I tried both versions and on my machine `parse_example` was slightly faster. Another thought was to do `dataset = dataset.batch(...)` before the parsing. Although, I haven't tested it so can't say if it is better.

If it is easier to read via `tf.parse_single_example` I'll happily revert the change. After all the focus here is readability over performance.

Reply:
I'd prefer the `parse_single_example` version for readability, please.

(I am curious at what the cause of different performance might be... applying `tf.parse_example()` to a vector containing one example follows almost exactly the same code path as applying `tf.parse_single_example()` to a scalar, except it does all of the necessary squeezing etc. internally. The way it deals with `tf.SparseTensor` appears to be [a bit clumsy](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/parsing_ops.py#L839-L847), but that path isn't used here, and apart from that I don't see any obvious deficiencies....)

Reply:
Okay, I've just moved batching and shuffling up to try decoding examples as batches rather than single examples. That does make a huge difference in speed and removes both reshape ops. I will still revert the changes and update the PR.

(Just FYI here are the times on my super recent GTX750Ti)

Decoding `parse_single_example`
    Step 0: loss = 2.31 (0.609 sec)
    Step 100: loss = 2.15 (0.031 sec)
    Step 200: loss = 1.96 (0.030 sec)
    Step 300: loss = 1.60 (0.029 sec)
    Step 400: loss = 1.35 (0.029 sec)
    Step 500: loss = 0.90 (0.031 sec)
    Step 600: loss = 0.88 (0.030 sec)
    Step 700: loss = 0.72 (0.031 sec)
    Step 800: loss = 0.62 (0.028 sec)
    Step 900: loss = 0.60 (0.031 sec)
    Step 1000: loss = 0.56 (0.027 sec)
    Done training for 2 epochs, 1100 steps.


Decoding `parse_example` (one example at a time):

    Step 0: loss = 2.30 (0.585 sec)
    Step 100: loss = 2.18 (0.031 sec)
    Step 200: loss = 1.99 (0.031 sec)
    Step 300: loss = 1.70 (0.031 sec)
    Step 400: loss = 1.39 (0.030 sec)
    Step 500: loss = 1.14 (0.029 sec)
    Step 600: loss = 0.73 (0.031 sec)
    Step 700: loss = 0.73 (0.028 sec)
    Step 800: loss = 0.75 (0.030 sec)
    Step 900: loss = 0.59 (0.028 sec)
    Step 1000: loss = 0.80 (0.030 sec)
    Done training for 2 epochs, 1100 steps.

Decoding `parse_example` batched:

    Step 0: loss = 2.35 (0.237 sec)
    Step 100: loss = 2.10 (0.004 sec)
    Step 200: loss = 1.79 (0.004 sec)
    Step 300: loss = 1.42 (0.004 sec)
    Step 400: loss = 1.14 (0.005 sec)
    Step 500: loss = 0.87 (0.003 sec)
    Step 600: loss = 0.64 (0.004 sec)
    Step 700: loss = 0.69 (0.004 sec)
    Step 800: loss = 0.75 (0.005 sec)
    Step 900: loss = 0.45 (0.004 sec)
    Step 1000: loss = 0.57 (0.005 sec)
    Done training for 2 epochs, 1100 steps.

Reply:
Wow, those numbers are concerning. It looks like there's a decent amount of low-hanging performance to be picked up in the `tf.parse_single_example()` implementation, so I'm going to prioritize that. 

Thanks for making the changes!